core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -1659,40 +1660,58 @@ sealed trait ZIO[-R, +E, +A]
   final def retry[R1 <: R, S](
     policy: => Schedule[R1, E, S]
   )(implicit ev: CanFail[E], trace: Trace): ZIO[R1, E, A] =
-    ZIO.suspendSucceed {
-
-      def loop(driver: Schedule.Driver[Any, R1, E, S]): ZIO[R1, E, A] =
-        self.catchAllCause { cause =>
-          cause.failureOrCause.fold(
-            e =>
-              driver
-                .next(e)
-                .foldZIO(
-                  _ => driver.last.orDie.flatMap(_ => Exit.failCause(cause)),
-                  _ => loop(driver)
-                ),
-            cause => Exit.failCause(cause)
+    policy.driver.flatMap { driver =>
```

Comment:
I'm not sure if it is worth using `whileLoop` for this, and retryN, and retryUntilZIO, and maybe repeatN